// @upscalerjs/core/dist/cjs/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidRange = exports.isNumber = exports.hasValidChannels = exports.isValidModelDefinition = exports.ModelDefinitionValidationError = exports.MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE = exports.isValidModelType = exports.isString = exports.isTensor = exports.isThreeDimensionalTensor = exports.isFourDimensionalTensor = exports.makeIsNDimensionalTensor = exports.isShape4D = void 0;
var tf = __importStar(require("@tensorflow/tfjs-core"));
var isShape4D = function (shape) {
    if (!Boolean(shape) || !Array.isArray(shape) || shape.length !== 4) {
        return false;
    }
    for (var _i = 0, shape_1 = shape; _i < shape_1.length; _i++) {
        var val = shape_1[_i];
        if (val !== null && typeof val !== 'number') {
            return false;
        }
    }
    return true;
};
exports.isShape4D = isShape4D;
function makeIsNDimensionalTensor(rank) {
    function fn(pixels) {
        try {
            return pixels.shape.length === rank;
        }
        catch (err) { }
        return false;
    }
    return fn;
}
exports.makeIsNDimensionalTensor = makeIsNDimensionalTensor;
exports.isFourDimensionalTensor = makeIsNDimensionalTensor(4);
exports.isThreeDimensionalTensor = makeIsNDimensionalTensor(3);
var isTensor = function (input) { return input instanceof tf.Tensor; };
exports.isTensor = isTensor;
var isString = function (el) { return typeof el === 'string'; };
exports.isString = isString;
var isValidModelType = function (modelType) { return typeof modelType === 'string' && ['layers', 'graph',].includes(modelType); };
exports.isValidModelType = isValidModelType;
var MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE;
(function (MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE) {
    MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE["UNDEFINED"] = "undefined";
    MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE["INVALID_MODEL_TYPE"] = "invalidModelType";
    MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE["MISSING_PATH"] = "missingPath";
})(MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE = exports.MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE || (exports.MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE = {}));
var ModelDefinitionValidationError = /** @class */ (function (_super) {
    __extends(ModelDefinitionValidationError, _super);
    function ModelDefinitionValidationError(type) {
        var _this = _super.call(this, type) || this;
        _this.type = type;
        return _this;
    }
    return ModelDefinitionValidationError;
}(Error));
exports.ModelDefinitionValidationError = ModelDefinitionValidationError;
var isValidModelDefinition = function (modelDefinition) {
    if (modelDefinition === undefined) {
        throw new ModelDefinitionValidationError(MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.UNDEFINED);
    }
    if (!(0, exports.isValidModelType)(modelDefinition.modelType || 'layers')) {
        throw new ModelDefinitionValidationError(MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.INVALID_MODEL_TYPE);
    }
    if (!modelDefinition.path) {
        throw new ModelDefinitionValidationError(MODEL_DEFINITION_VALIDATION_CHECK_ERROR_TYPE.MISSING_PATH);
    }
    return true;
};
exports.isValidModelDefinition = isValidModelDefinition;
var hasValidChannels = function (tensor) { return tensor.shape.slice(-1)[0] === 3; };
exports.hasValidChannels = hasValidChannels;
var isNumber = function (el) { return typeof el === 'number'; };
exports.isNumber = isNumber;
var isValidRange = function (range) { return Array.isArray(range) && range.length === 2 && range.every(exports.isNumber); };
exports.isValidRange = isValidRange;
